domain/cloudimagemetadata/state/state.go

```diff
@@ -0,0 +1,381 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/cloudimagemetadata"
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/logger"
+	"github.com/juju/juju/domain"
+	cloudmetadataerrors "github.com/juju/juju/domain/cloudimagemetadata/errors"
+	dberrors "github.com/juju/juju/internal/database"
+)
+
+const (
+	// CustomSource defines a metadata as "custom" meaning it won't expires
+	CustomSource = "custom"
+	// ExpirationDelay is the maximum time a metadata can live in the cache before being removed.
+	ExpirationDelay = 5 * time.Minute
+)
+
+// State encapsulates the state management, logging, and architecture data.
+type State struct {
+	*domain.StateBase
+	logger                logger.Logger
+	architectureIDsByName map[string]int
+	archesOnce            sync.Once
+}
+
+// NewState creates a new State instance using the provided database transaction factory and logger.
+func NewState(factory database.TxnRunnerFactory, logger logger.Logger) *State {
+	state := &State{
+		StateBase: domain.NewStateBase(factory),
+		logger:    logger,
+	}
+	return state
+}
+
+// SaveMetadata stores the provided list of cloud image metadata into the database.
+// Returns [cloudmetadataerrors.Invalid] if the provided metadata omits required fields.
+// Returns [cloudmetadataerrors.MultipleInsertNonUnique] if several metadatas passed as parameters
+// overrides each other.
+// It also fires a cleanup for old images, if any.
+func (s *State) SaveMetadata(ctx context.Context, metadatas []cloudimagemetadata.Metadata) error {
+	// Cleaning the database before insert/update is a reasonable way to remove outdated data without
+	// requiring an housekeeping goroutine. Since a metadata data doesn't live long in the db and
+	// it shouldn't be a lot of custom metadata, it shouldn't slow too much the insertion.
+	// Expired metadata are ignored when retrieving metadata, so it doesn't have any impact to keep
+	// some of them longer than their time to live.
+	if err := s.tryCleanUpExpiredMetadata(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	if err := s.ensuresArchitecturesLoaded(ctx); err != nil {
```

Comment:
The transaction inside `ensuresArchitecturesLoaded()` will be separate from the main tx in `SaveMetadata()`. Is this wanted? maybe those queries should be part of one single transaction?

Reply:
If the reason is that `sync.Once` then maybe we should lift some of that logic to the service.

Reply:
This method is just to avoid a complicated insert method with either a temp table or a `WITH` clause to gather information for "fresh" architecture table.

SQLAIR doesn't support complex INSERT with WITH clause out of the box (it get confused).

So, since `architecture` table is constant and short (and doesn't need to become variable neither big)
https://github.com/juju/juju/blob/fde60554d4f52d2b6ad326d20894aa8ea3e4f3c8/domain/schema/controller/sql/0021-platform.sql#L1-L14

I think it is simpler to load it once in a local map to prepare input data (convert arch from name to id), and insert them easily).

Loading this table could be done every time we try to save metadata, but it quite easy to avoid it trough this code.

It shouldn't be leaked to the service, because it is purely implementation detail (in the Find method, id to name are done by a JOIN).

Reply:
I understand why you are doing it, and your limitation seems to come from sqlair. Nonetheless IMO this `SaveMetadata` method is doing too much: 
- It potentially runs 3 transactions (it will *at least* run 2 transactions), which is more expensive than adding another query to a single transaction.
- Mixes cleanup and insertion logic. 

But my main issue is with the architectures cache. I see our ddl as the source of truth of our entire modelling. Accessing and mutating data should use the same semantics that define the model, i.e. SQL. Even for very simple lookup constant table as this one. 
Ideally I'd love to see one single batch insert SQL query used in this method. 

Anyways, I don't want to block this PR, I'd like some inputs from @SimonRichardson and @manadart on this subject though.

Reply:
@SimonRichardson ask to get ride of this one and just hard code architecture map and add a TU to check that hardcoded and db values don't diverge.

This will remove the (rare) transaction for loading arch.

However, the one which delete expired items can't be removed. Another approach would to have a goroutine which take care of removing oldest image, but I think it would be worse on the part of resources consumption (image are not added that frequently).
Moreover, it has no sense to be but in the same transaction than adding values, since both are functionally unrelated.

I had though about a way to put this transaction less frequent (through a semaphor which fire the transaction only if there is none actually running, or even none actually running since expiration delay), but it feels like premature optimization.

Anyhow `SaveMetadata` will requires to be updated regarding other requests from Simon.

Reply:
Regarding the separate goroutine vs the call during `SaveMetadata` I'm ok with both approaches. But if you choose to delete **in** `SaveMetadata` then please put the SQL in a separate method and orchestrate the two calls on `SaveMetadata` at the service layer. 
I would really like single-transaction methods on the state layer, and orchestration on the service layer if needed.

Reply:
As discussed with the team, we are OK with keeping the two transactions on this same method, but keeping in mind that the first transaction (cleanup) should not make the `SaveMetadata` method fail if it fails, so the error should only be logged.